gapir/cc/stack.h

```diff
@@ -26,457 +26,188 @@
 #include <stdint.h>
 #include <string.h>
 
+#include <assert.h>
 #include <type_traits>
-#include <vector>
+#include <typeindex>
 
 namespace gapir {
 
-// Strongly typed, limited size stack for the stack based virtual machine. If an
-// invalid operation is called on the stack then the stack will go into an
-// invalid state where each operation is no op (except print stack). From an
-// invalid state the stack can't go back to a valid state again.
 class Stack {
  public:
-  // Representation of an unconverted value from the stack.
-  typedef uint64_t BaseValue;
-
-  // Construct a new stack with the given size and memory manager
-  // The memory manager is needed to resolve constant and volatile pointers to
-  // absolute pointers
-  Stack(uint32_t size, const MemoryManager* memoryManager);
-
-  // Pop the element from the top of the stack as the given type if its type is
-  // matching with the given type. For pointer types convert the pointer to an
-  // absolute pointer before returning it. Puts the stack into an invalid state
-  // if called on an empty stack or if the requested type doesn't match with the
-  // type of the value at the top of the stack.
+  class VolatilePointer {
+   public:
+    explicit VolatilePointer(uint32_t offset) : mOffset(offset) {}
+    bool operator==(const VolatilePointer& rhs) const {
+      return mOffset == rhs.mOffset;
+    }
+    uint32_t getOffset() const { return mOffset; }
+
+   private:
+    uint32_t mOffset;
+  };
+
+  class ConstantPointer {
+   public:
+    explicit ConstantPointer(uint32_t offset) : mOffset(offset) {}
+    bool operator==(const ConstantPointer& rhs) const {
+      return mOffset == rhs.mOffset;
+    }
+    uint32_t getOffset() const { return mOffset; }
+
+   private:
+    uint32_t mOffset;
+  };
+
+  Stack(uint32_t size, const MemoryManager* memoryManager)
+      : mNumEntries(0),
+        mMaxEntries(size),
+        mStack(new Entry[size]),
+        mMemoryManager(memoryManager) {}
+
+  ~Stack() { delete[] mStack; }
+
   template <typename T>
-  T pop() {
-    if (!popCheck("pop")) {
-      return T();
+  T pop(typename std::enable_if<std::is_pointer<T>::value>::type* = 0) {
```

Comment:
Please add method comments here (Looks like most of the code-removal was comments)